### ISSUE TITLE
Data explorer accessibility issues (analysis)

### ISSUE THREAD

**Issue by nightsh** [CivicActions/ckanext-ed#225]
The CKAN data explorer is not 508 compliant and lacks a few essential features, such as keyboard only navigation and elements metadata to help screen readers identify the contents of the widget.

I've been investigating this in the past days. Here is a write up of my findings so far.

CKAN uses [Recline](https://github.com/okfn/recline) to display the preview of a selected resource. Currently, the Recline version we're using lacks the ability to draw tables that are accessible for screen readers. Some of the critical features it lacks are:

- the (column) headings are now selectable via keyboard (so the user does not know what the values represent)
- IF the column headings would have been selectable, the context menu is only triggered by a mouse right click (that menu toggles on/off columns for display)
- focusing order (e.g. tabindex HTML attribute) does not feel natural and does not follow user's input. A solution would be to specifically trigger focus events according to such actions, such as focusing the table cells right after applying filters.

However, Recline has a dependency of its own, called [SlickGrid](http://slickgrid.net/). This is a lightweight JS library for drawing tables, but its official version is unmaintained for a while now. There is a fork though, and I have found some interest in fixing a few 508 issues:

- https://github.com/6pac/SlickGrid/issues/87
- https://github.com/6pac/SlickGrid/issues/206

Trying to apply those suggestions fails in our case, because the SlickGrid versions those patches apply to are newer than ours. Trying to upgrade SlickGrid in Recline breaks it because of two other dependency problems:

- newer SlickGrid depends on a newer jQuery version
- incompatible usage (newer versions are not backwards compatible)

One potential solution would be to backport such fixes to the SlickGrid version we currently use, while keeping jQuery and Recline versions as they are. This might work, as they're mostly template and events changes. However, this takes time and would likely break compatibility with existing implementation, so we would need to fork it and maintain it.

Another way of doing this is upgrading everything. Trying that broke pretty much the entire UI, proving that it's a much slower process than upgrading everything at once and fixing the errors.

---

Map tiles attributes.

Another feature of the data explorer (through Recline) is displaying a map. The map, however, is not accessible for screen readers as the map tiles don't have the necessary ARIA attributes.

The library (another 3rd party dependency) doing this is called [Leaflet](https://leafletjs.com/). Newer versions of Leaflet claim to have improved support for screen readers, but I haven't been able to load it from Ed without breaking Recline. If there's a way to completely override a JS lib loaded from another CKAN extension, then this issue might be very easy to fix.

So far, the error I was getting while trying to upgrade Leaflet through ckanext-ed was that the required module was already loaded. I will attach the error message once I reproduce it again, as of now I don't have that code around anymore.

---

Conclusions:

- it's possible to fix the table view by upgrading the libraries and fixing compatibility issues between versions (high risk of breaking stuff, slow and complicated process)
- it's possible to fix the table view by patching SlickGrid and Recline (medium-high risk, still a slow process, requires deeper understanding of how these JS libraries are working)
- we can probably fix the Leaflet map tiles by upgrading Leaflet, we just need to find a way to load it without breaking CKAN's assets logic
